File: testbed_ruby/test/associations/optional_one_to_one_test.rb
require 'test/unit'
require 'src/student_b'
require 'src/mentor_b'
require 'src/program_b'

module CruiseTest
class OptionalOneToOneTest < Test::Unit::TestCase

  def test_Constructor
    student = StudentB.new
    mentor = MentorB.new(student)
    
    assert_equal(student,mentor.get_student)
    assert_equal(mentor,student.get_mentor)
  end 
  
  def test_CannotSetStudentBToNull
    student = StudentB.new
    mentor = MentorB.new(student)
    assert_equal(false,mentor.set_student(nil))
  end

  def test_SetStudent
    student = StudentB.new
    mentor = MentorB.new(student)
    
    student2 = StudentB.new
    assert_equal(true,mentor.set_student(student2))
    
    
    assert_equal(mentor,student2.get_mentor)
    assert_equal(student2,mentor.get_student)
    assert_equal(nil,student.get_mentor)
  end 
  
  def test_SetMentorCannotReset
    s = StudentB.new
    m = MentorB.new(s)
    
    s2 = StudentB.new
    m2 = MentorB.new(s2)
    
    assert_equal(false,m.set_student(s2))
    assert_equal(m,s.get_mentor)
    assert_equal(s,m.get_student)
    assert_equal(m2,s2.get_mentor)
    assert_equal(s2,m2.get_student)
  end 

  def test_SetMentor
    student = StudentB.new
    mentor = MentorB.new(student)
    
    student2 = StudentB.new
    student2.set_mentor(mentor)
    
    assert_equal(mentor,student2.get_mentor)
    assert_equal(student2,mentor.get_student)
    assert_equal(nil,student.get_mentor)
  end   

  def test_UnableToConstructNewSubordinateFromExistingDriverThatAlreadyHasDriver
  	assert_raise RuntimeError do #"Unable to create mentor due to student"
      student = StudentB.new
      MentorB.new(student)
      MentorB.new(student) 
    end
  end 


  def test_DeleteDriverStudentHasNoMentor
    student = StudentB.new
    student.delete
    assert_equal(nil,student.get_mentor)
  end
  
  def test_DeleteDriverRemovesSubordinate
    student = StudentB.new
    mentor = MentorB.new(student)
    program = ProgramB.new
    mentor.set_program(program)
    
    student.delete
    assert_equal(nil,student.get_mentor)
    assert_equal(nil,mentor.get_student)
    assert_equal(nil,mentor.get_program)
  end  
  
  def test_DeleteSubordinateKeepDriver
    student = StudentB.new
    mentor = MentorB.new(student)
    program = ProgramB.new
    student.set_program(program)
    
    mentor.delete
    assert_equal(nil,student.get_mentor)
    assert_equal(nil,mentor.get_student)
    assert_equal(program,student.get_program)
  end  
  
end
end

File: testbed_php/test/associations/OptionalOneToOneTest.php
<?php

class OptionalOneToOneTest extends UnitTestCase
{


  public function test_Constructor()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    
    $this->assertEqual($student,$mentor->getStudent());
    $this->assertEqual($mentor,$student->getMentor());
  } 
  
  public function test_CannotSetStudentBToNull()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    $this->assertEqual(false,$mentor->setStudent(null));
  }
    

  public function test_SetStudent()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    
    $student2 = new StudentB();
    $this->assertEqual(true,$mentor->setStudent($student2));
    
    
    $this->assertEqual($mentor,$student2->getMentor());
    $this->assertEqual($student2,$mentor->getStudent());
    $this->assertEqual(null,$student->getMentor());
  } 
  
  public function test_SetMentorCannotReset()
  {
    $s = new StudentB();
    $m = new MentorB($s);
    
    $s2 = new StudentB();
    $m2 = new MentorB($s2);
    
    $this->assertEqual(false,$m->setStudent($s2));
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($s,$m->getStudent());
    $this->assertEqual($m2,$s2->getMentor());
    $this->assertEqual($s2,$m2->getStudent());
  } 

  

  public function test_SetMentor()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    
    $student2 = new StudentB();
    $student2->setMentor($mentor);
    
    $this->assertEqual($mentor,$student2->getMentor());
    $this->assertEqual($student2,$mentor->getStudent());
    $this->assertEqual(null,$student->getMentor());
  }   

  public function test_UnableToConstructNewSubordinateFromExistingDriverThatAlreadyHasDriver()
  {
  	$this->expectException(new Exception("Unable to create mentor due to student"));
    $student = new StudentB();
    new MentorB($student);
    new MentorB($student); 
  } 


  public function test_DeleteDriverStudentHasNoMentor()
  {
    $student = new StudentB();
    $student->delete();
    $this->assertEqual(null,$student->getMentor());
  }
  

  public function test_DeleteDriverRemovesSubordinate()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    $program = new ProgramB();
    $mentor->setProgram($program);
    
    $student->delete();
    $this->assertEqual(null,$student->getMentor());
    $this->assertEqual(null,$mentor->getStudent());
    $this->assertEqual(null,$mentor->getProgram());
  }  
  

  public function test_DeleteSubordinateKeepDriver()
  {
    $student = new StudentB();
    $mentor = new MentorB($student);
    $program = new ProgramB();
    $student->setProgram($program);
    
    $mentor->delete();
    $this->assertEqual(null,$student->getMentor());
    $this->assertEqual(null,$mentor->getStudent());
    $this->assertEqual($program,$student->getProgram());
  }  
  
}
?>